### DISCUSSION TITLE
OpenURL action crash on macOS Sequoia

### DISCUSSION THREAD

**Discussion by beetee17** [pointfreeco/swift-dependencies#301] [Q&A]
Hi,

I've been using the default live implementation of `openURL` in a TCA production app, and have received multiple crash reports from users that point to an issue with `openURL` when calling it from an effect like so: `.run { _ in await openURL(...) }`.

It seems that the code is not running on the `MainActor`, even though the `Task` is marked `@MainActor`, causing a dispatch queue assertion failure. Weirdly, this assertion is not triggered on my machine that runs macOS Sonoma.

The crash reports are of the following format:

```
Thread 7 Crashed:
0   libdispatch.dylib             	0x0000000190c079c0 _dispatch_assert_queue_fail + 120 (queue.c:54)
1   libdispatch.dylib             	0x0000000190c07948 dispatch_assert_queue + 196 (queue.c:84)
2   libswift_Concurrency.dylib    	0x0000000271cd6e08 swift_task_isCurrentExecutorImpl(swift::SerialExecutorRef) + 284 (Actor.cpp:0)
3   RedactedApp                	0x00000001078f59b4 closure #1 in closure #1 in closure #1 in closure #1 in variable initialization expression of static OpenURLKey.liveValue + 92 (OpenURL.swift:0)
4   SwiftUI                       	0x00000001c06864e8 partial apply for closure #1 in closure #1 in SwiftUIGlue.defaultOpenSensitiveURLAction() + 36
5   SwiftUI                       	0x00000001c066ede0 thunk for @escaping @callee_guaranteed (@guaranteed NSRunningApplication?, @guaranteed Error?) -> () + 88 (<compiler-generated>:0)
6   AppKit                        	0x00000001955120ac _NSWorkspaceHandleLSOpenResult + 156 (NSWorkspace.m:1379)
7   LaunchServices                	0x000000019148d3ac _LSFireCompletionHandler(void (__LSASN const*, unsigned char, __CFError*) block_pointer, NSURL*, NSURL*, NSDictionary<NSString*, objc_object*>*, objc_object*, bool, NSError*) + 200 (LSOpenNewWorld.mm:1220)
8   LaunchServices                	0x00000001914915a4 _LSOpenAsyncCommonDispatched(objc_object*, NSArray<NSURL*>*, NSURL*, objc_object*, NSDictionary<NSString*, objc_object*>*, NSObject<OS_dispatch_semaphore>*, void (__LSASN const*, unsigned char, __C... + 2560 (LSOpenNewWorld.mm:1394)
9   LaunchServices                	0x0000000191491b58 invocation function for block in _LSOpenAsyncCommon(objc_object*, NSArray<NSURL*>*, NSURL*, objc_object*, NSDictionary<NSString*, objc_object*>*, void (__LSASN const*, unsigned char, __CFError*) bl... + 352 (LSOpenNewWorld.mm:1275)
10  libdispatch.dylib             	0x0000000190c03854 _dispatch_call_block_and_release + 32 (init.c:1549)
11  libdispatch.dylib             	0x0000000190c055b4 _dispatch_client_callout + 20 (object.m:576)
12  libdispatch.dylib             	0x0000000190c0cbd8 _dispatch_lane_serial_drain + 768 (queue.c:3934)
13  libdispatch.dylib             	0x0000000190c0d764 _dispatch_lane_invoke + 432 (queue.c:4025)
14  libdispatch.dylib             	0x0000000190c189a0 _dispatch_root_queue_drain_deferred_wlh + 288 (queue.c:7193)
15  libdispatch.dylib             	0x0000000190c181ec _dispatch_workloop_worker_thread + 540 (queue.c:6787)
16  libsystem_pthread.dylib       	0x0000000190db43d8 _pthread_wqthread + 288 (pthread.c:2696)
17  libsystem_pthread.dylib       	0x0000000190db30f0 start_wqthread + 8
```

I was able to create and test a fix for this crash by overriding the dependency with an implementation of `openURL` that marks the outermost closure as `@MainActor`:

```swift
OpenURLEffect { @MainActor url in
	let stream = AsyncStream<Bool> { continuation in
		let task = Task { @MainActor in
			#if os(watchOS)
			EnvironmentValues().openURL(url)
			continuation.yield(true)
			continuation.finish()
			#else
			EnvironmentValues().openURL(url) { canOpen in
				continuation.yield(canOpen)
				continuation.finish()
			}
			#endif
		}
		continuation.onTermination = { @Sendable _ in
			task.cancel()
		}
	}
	return await stream.first(where: { _ in true }) ?? false
}
```

### Dependencies version information

1.4.1

### Destination operating system

macOS Sequoia 15.0

### Xcode version information

XCode 16.0

### Swift Compiler version information

6.0

P.S. Did not want to open an issue as it is hard for me to test on macOS Sequoia, so I could not do many checks.

**Comment by stephencelis** [comment]
@beetee17 Thanks for the report! Are you able to open a PR with your fix and a test? I'd be happy to run it for you on Sequoia to check it.

**Reply by beetee17** [reply]
@stephencelis I have created a pull request with the changes that fixed the crash for my users (#305). Though this could be an integration bug, as the app calls `openURL` from a TCA reducer's `.run` effect. 

FWIW, the exact fix I tested was to override the problematic reducer's dependency with `.dependency(\.openURL, OpenURLEffect { @MainActor ... })`, rather than modifying the `liveValue` implementation directly (which is what #305 is doing). Not sure if this makes any difference.

EDIT: I just got reports from users who were able to test an alternative fix that I tried, which was to write the reducer like so: `.run { @MainActor _ in await openURL(...) }`, with no modifications to the dependency. This does not work and the crash persists.
